src/lib.rs

```diff
@@ -124,171 +124,401 @@ use proc_macro::TokenStream;
+
+	let mod_code = if args.erased {
+		let lifetimes = trait_generics.lifetimes();
+		let const_params = trait_generics.const_params();
+		let type_params = trait_generics
+			.type_params()
+			.map(|syn::TypeParam { ident, .. }| -> syn::TypeParam {
+				parse_quote!( #ident : ?Sized )
+			});
+
+		// Token here is used as an argument to a function in order to seal it.
+		quote! {
+			pub trait Sealed< #(#lifetimes ,)* #(#type_params ,)* #(#const_params ,)* > {}
+			pub struct Token;
+		}
+	} else {
+		// `trait_generics` does not output its where clause when tokenized (due
+		// to supertraits in the middle). So we output them separately.
+		quote! {
+			use super::*;
+			pub trait Sealed #trait_generics #where_clause {}
+			pub struct Token;
+		}
+	};
+
+	quote! {
+		#[automatically_derived]
+		#vis mod #seal {
+			#mod_code
+		}
+		#item_trait
+	}
 }
 
-fn parse_sealed_impl(item_impl: &syn::ItemImpl) -> syn::Result<TokenStream2> {
-    let impl_trait = item_impl
-        .trait_
-        .as_ref()
-        .ok_or_else(|| syn::Error::new_spanned(item_impl, "missing implementation trait"))?;
+fn parse_sealed_impl(item_impl: &mut syn::ItemImpl) -> syn::Result<TokenStream2> {
+	let impl_trait = item_impl
+		.trait_
+		.as_ref()
+		.ok_or_else(|| {
+			syn::Error::new_spanned(item_impl.clone(), "missing implementation trait")
+		})?;
 
-    let mut sealed_path = impl_trait.1.segments.clone();
+	let mut sealed_path = impl_trait.1.segments.clone();
 
-    // since `impl for ...` is not allowed, this path will *always* have at least length 1
-    // thus both `first` and `last` are safe to unwrap
-    let syn::PathSegment { ident, arguments } = sealed_path.pop().unwrap().into_value();
-    let seal = seal_name(ident.unraw());
-    sealed_path.push(parse_quote!( #seal ));
-    sealed_path.push(parse_quote!(Sealed));
+	// since `impl for ...` is not allowed, this path will *always* have at least length 1
+	// thus both `first` and `last` are safe to unwrap
+	let syn::PathSegment { ident, arguments } = sealed_path
+		.pop()
+		.unwrap()
+		.into_value();
+	let seal = seal_name(ident.unraw());
+	sealed_path.push(parse_quote!( #seal ));
+	sealed_path.push(parse_quote!(Sealed));
 
-    let self_type = &item_impl.self_ty;
+	let self_type = &item_impl.self_ty;
 
-    // Only keep the introduced params (no bounds), since
-    // the bounds may break in the `#seal` submodule.
-    let (trait_generics, _, where_clauses) = item_impl.generics.split_for_impl();
+	// Only keep the introduced params (no bounds), since
+	// the bounds may break in the `#seal` submodule.
+	let (trait_generics, _, where_clauses) = item_impl
+		.generics
+		.split_for_impl();
 
-    Ok(quote! {
-        #[automatically_derived]
-        impl #trait_generics #sealed_path #arguments for #self_type #where_clauses {}
-        #item_impl
-    })
+	for item in &mut item_impl.items {
+		if let syn::ImplItem::Fn(item) = item {
+			let input = parse_function_arguments(&item.attrs)?;
```

Comment:
Can we: 
- make `FunctionArguments` consider the whole attribute (i.e. `#[seal(...)]`
- get the respective token stream and use `parse_macro_input!` 

That way we get, less code and more concise code